deselfify doesn't work with generic types

Methods using `Self` in their signatures should be mockable because the `deselfify` routine replaces `Self` with the mock type.  However, it fails to consider type parameters.  Hence, the following code doesn't work:
```rust
#[automock]
trait Foo<T: 'static> {
    fn new(t: T) -> Self;
}
```
```
error[E0107]: wrong number of type arguments: expected 1, found 0
 --> mockall/tests/automock_constructor_in_generic_trait.rs:7:7
  |
7 | trait Foo<T: 'static> {
  |       ^^^ expected 1 type argument

error: aborting due to previous error
```
Note that such methods can still be mocked using `mock!` instead of `#[automock]`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Fixed
 
+- Methods of generics structs and traits that reference `Self` (such as
+  constructors or comparators) can now be mocked without any hacks.
+  ([#20](https://github.com/asomers/mockall/pull/20))
+
 - Specializing methods of generic structs (and traits) can now be mocked
   without the hack of duplicating the struct's (or trait's) generic parameters.
   ([#19](https://github.com/asomers/mockall/pull/19))

--- a/mockall/tests/automock_constructor_in_generic_trait.rs
+++ b/mockall/tests/automock_constructor_in_generic_trait.rs
@@ -1,0 +1,19 @@
+// vim: tw=80
+//! A generic trait with a non-generic constructor method.
+
+use mockall::*;
+
+#[automock]
+trait Foo<T: 'static> {
+    fn new(t: T) -> Self;
+}
+
+#[test]
+fn return_once() {
+    let mock = MockFoo::<u32>::default();
+
+    MockFoo::<u32>::expect_new()
+        .return_once(move |_| mock);
+
+    let _mock = MockFoo::new(5u32);
+}

--- a/mockall_derive/src/expectation.rs
+++ b/mockall_derive/src/expectation.rs
@@ -719,7 +719,7 @@ pub(crate) fn expectation(attrs: &TokenStream,
             let mut rt = ty.clone();
             destrify(&mut rt);
             if let Some(i) = self_ident {
-                crate::deselfify(&mut rt, i);
+                crate::deselfify(&mut rt, i, generics);
             }
             if let Type::Reference(ref tr) = rt.as_ref() {
                 if tr.lifetime.as_ref().map_or(false, |lt| lt.ident == "static")

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -218,25 +218,26 @@ fn demutify_arg(arg: &mut ArgCaptured) {
     };
 }
 
-/// Replace any references to `Self` in `literal_type` with `actual`.  Useful
-/// for constructor methods
-fn deselfify(literal_type: &mut Type, actual: &Ident) {
+/// Replace any references to `Self` in `literal_type` with `actual`.
+/// `generics` is the Generics field of the parent struct.  Useful for
+/// constructor methods.
+fn deselfify(literal_type: &mut Type, actual: &Ident, generics: &Generics) {
     match literal_type {
         Type::Slice(s) => {
-            deselfify(s.elem.as_mut(), actual);
+            deselfify(s.elem.as_mut(), actual, generics);
         },
         Type::Array(a) => {
-            deselfify(a.elem.as_mut(), actual);
+            deselfify(a.elem.as_mut(), actual, generics);
         },
         Type::Ptr(p) => {
-            deselfify(p.elem.as_mut(), actual);
+            deselfify(p.elem.as_mut(), actual, generics);
         },
         Type::Reference(r) => {
-            deselfify(r.elem.as_mut(), actual);
+            deselfify(r.elem.as_mut(), actual, generics);
         },
         Type::Tuple(tuple) => {
             for elem in tuple.elems.iter_mut() {
-                deselfify(elem, actual);
+                deselfify(elem, actual, generics);
             }
         }
         Type::Path(type_path) => {
@@ -246,16 +247,46 @@ fn deselfify(literal_type: &mut Type, actual: &Ident) {
             let p = &mut type_path.path;
             for seg in p.segments.iter_mut() {
                 if seg.ident == "Self" {
-                    seg.ident = actual.clone()
+                    seg.ident = actual.clone();
+                    if let PathArguments::None = seg.arguments {
+                        //let (_, tg, _) = generics.split_for_impl();
+                        if !generics.params.is_empty() {
+                            let args = Punctuated::from_iter(
+                                generics.params.iter().map(|gp| {
+                                    let ident = match gp {
+                                        GenericParam::Type(tp) => &tp.ident,
+                                        GenericParam::Lifetime(ld) =>
+                                            &ld.lifetime.ident,
+                                        GenericParam::Const(cp) => &cp.ident,
+                                    };
+                                    parse2::<GenericArgument>(
+                                        quote!(#ident)
+                                    ).unwrap()
+                                })
+                            );
+                            seg.arguments = PathArguments::AngleBracketed(
+                                AngleBracketedGenericArguments {
+                                    colon2_token: None,
+                                    lt_token: generics.lt_token.unwrap().clone(),
+                                    args,
+                                    gt_token: generics.gt_token.unwrap().clone(),
+
+                                }
+                            );
+                        }
+                    } else {
+                        compile_error(seg.arguments.span(),
+                            "Type arguments after Self are unexpected");
+                    }
                 }
                 if let PathArguments::AngleBracketed(abga) = &mut seg.arguments
                 {
                     for arg in abga.args.iter_mut() {
                         match arg {
                             GenericArgument::Type(ty) =>
-                                deselfify(ty, actual),
+                                deselfify(ty, actual, generics),
                             GenericArgument::Binding(b) =>
-                                deselfify(&mut b.ty, actual),
+                                deselfify(&mut b.ty, actual, generics),
                             _ => /* Nothing to do */(),
                         }
                     }
@@ -263,10 +294,10 @@ fn deselfify(literal_type: &mut Type, actual: &Ident) {
             }
         },
         Type::Paren(p) => {
-            deselfify(p.elem.as_mut(), actual);
+            deselfify(p.elem.as_mut(), actual, generics);
         },
         Type::Group(g) => {
-            deselfify(g.elem.as_mut(), actual);
+            deselfify(g.elem.as_mut(), actual, generics);
         },
         Type::Macro(_) | Type::Verbatim(_) => {
             compile_error(literal_type.span(),

